### PR TITLE
Feature/chart.js

### PR DESCRIPTION
**Pull request by yesod30**
Remove the old wrapper c# library for chart js and replace it with the newset version of chartjs, while converting everything to use JS for chartjs itself.

### CODE REVIEW

File: WoWsShipBuilder.Web/Pages/Charts.razor [hunk 1]
```diff
@@ -40,24 +34,27 @@
...
-                        }
-                    </MudSelect>
-                </MudItem>
-                <MudItem Class="z-10" xs="12">
-                    <Chart Config="verticalDispConfig"/>
-                </MudItem>
-            </MudGrid>
-        </MudContainer>
+        <div>
+            <canvas id="@HorizontalDispersionId"></canvas>
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Pages/Charts.razor:38]: Is it possible to use @ref instead?

File: WoWsShipBuilder.Web/Pages/Charts.razor [hunk 2]
```diff
@@ -40,24 +34,27 @@
...
+                    {
+                        <MudSelectItem Value="@item">
+                            <MudText Color="Color.Secondary">@Localizer.GetAppLocalization(Translation.ResourceManager.GetString(item.ToString())!).Localization</MudText>
+                        </MudSelectItem>
+                    }
+                </MudSelect>
+            </MudItem>
+            <MudItem Class="z-10" xs="12">
+                <div>
+                    <canvas id="@VerticalDispersionId"></canvas>
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Pages/Charts.razor:53]: Same as before, why not @ref?

File: WoWsShipBuilder.Web/Pages/Charts.razor [hunk 3]
```diff
@@ -75,19 +72,19 @@
...
-                </MudItem>
-            </MudGrid>
-        </MudContainer>
+        <MudGrid Class="d-inline-flex" Style="flex-direction: column">
+            <MudItem Class="z-20 mt-1 ml-auto mr-5 d-inline-flex" xs="2" md="1" Style="position: absolute; align-self: flex-end">
+                <MudNumericField Value="range" Label="@Localizer.GetAppLocalization(nameof(Translation.ShipStats_Range)).Localization" Variant="Variant.Outlined" Immediate="true" ValueChanged="UpdateTrajectory" T="double" DebounceInterval="250" Disabled="processing" Margin="Margin.Dense" Min="0"/>
+            </MudItem>
+            <MudItem Class="z-10" xs="12">
+                <div>
+                    <canvas id="@TrajectoryId"></canvas>
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Pages/Charts.razor:81]: Same as before, why not @ref?

File: WoWsShipBuilder.Web/Pages/Charts.razor [hunk 4]
```diff
@@ -102,44 +99,48 @@
                 </MudText>
             }
         </MudButton>
-        <MudContainer>
-            <Chart Config="penetrationConfig"/>
-            <Chart Config="flightTimeConfig"/>
-            <Chart Config="impactVelocityConfig"/>
-            <Chart Config="impactAngleConfig"/>
-        </MudContainer>
+        <div>
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Pages/Charts.razor:102]: Same as before, why not @ref?

File: WoWsShipBuilder.Web/Pages/Charts.razor [hunk 5]
```diff
@@ -149,365 +150,310 @@
...
+                        return;
+                    }
+                    selectedShipList = ((IEnumerable<ShellSelectionWrapper>) result.Data).ToList();
+                }
+                await UpdateShipsBatch(selectedShipList);
+            }
         }
-        await UpdateCharts(((IEnumerable<ShellSelectionWrapper>) selectedShipList.Data).ToList());
-        processing = false;
+        SetProcessing(false);
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Pages/Charts.razor:238]: Move into if condition, otherwise this will trigger an infinite render loop

File: WoWsShipBuilder.Web/Pages/Charts.razor [hunk 6]
```diff
@@ -149,365 +150,310 @@
...
+            await ChangeSuggestedMax(TrajectoryId, range / 4 * 1000);
+        }
     }
-
-    private async Task UpdateCharts(List<ShellSelectionWrapper> selectedShips)
+    
+    private async Task UpdateShipsBatch(List<ShellSelectionWrapper> shipList)
     {
+        List<string> chartIds = GetChartsIdList();
+        List<List<IEnumerable<Point>>> data =  new();
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Pages/Charts.razor:270]: why this structure? Is a triple-nested collection really necessary here?

File: WoWsShipBuilder.Web/Pages/Charts.razor [hunk 7]
```diff
@@ -149,365 +150,310 @@
...
+            await module.InvokeVoidAsync("CreateChart", FlightTimeId, Localizer.GetAppLocalization(nameof(Translation.DispersionGraphWindow_FlightTime)).Localization, rangeString, Localizer.GetAppLocalization(nameof(Translation.DispersionGraphWindow_FlightTime)).Localization, km, s);
+            await module.InvokeVoidAsync("CreateChart", ImpactVelocityId, Localizer.GetAppLocalization(nameof(Translation.DispersionGraphWindow_ImpactVelocity)).Localization, rangeString, Localizer.GetAppLocalization(nameof(Translation.DispersionGraphWindow_ImpactVelocity)).Localization, km, mps);
+            await module.InvokeVoidAsync("CreateChart", ImpactAngleId, Localizer.GetAppLocalization(nameof(Translation.DispersionGraphWindow_ImpactAngle)).Localization, rangeString, Localizer.GetAppLocalization(nameof(Translation.DispersionGraphWindow_ImpactAngle)).Localization, km, degree);
+            await ChangeSuggestedMax(TrajectoryId, range / 4 * 1000);
+        }
     }
-
-    private async Task UpdateCharts(List<ShellSelectionWrapper> selectedShips)
+    
+    private async Task UpdateShipsBatch(List<ShellSelectionWrapper> shipList)
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Pages/Charts.razor:267]: Rename to UpdateShipsBatchAsync

File: WoWsShipBuilder.Web/Pages/Charts.razor [hunk 8]
```diff
@@ -149,365 +150,310 @@
...
+                await UpdateShipsBatch(selectedShipList);
+            }
         }
-        await UpdateCharts(((IEnumerable<ShellSelectionWrapper>) selectedShipList.Data).ToList());
-        processing = false;
+        SetProcessing(false);
     }
 
-    private async Task<DialogResult?> GetSelectedShips()
+    private async Task SetupCharts()
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Pages/Charts.razor:241]: Rename to SetupChartsAsync

File: WoWsShipBuilder.Web/Pages/Charts.razor [hunk 9]
```diff
@@ -149,365 +150,310 @@
...
+            vertDispersionsCache.Add(label, verticalDispersions);
         }
-        IEnumerable<Point> pointsList = validData.First().Value.Coordinates.Select(x => new Point(x.X / 1000, x.Y));
-        AddChartDataset(name, pointsList, trajectoryConfig, chartLineColor);
+        return ChartsHelper.SelectVerticalDispersionDataset(verticalDispersions, selectedVertDispPlane);
     }
-
-    private void AddBallisticChartDataset(string name, Dictionary<double, Ballistic> data, string chartLineColor)
+    
+    private async Task ChangeVerticalDispPlane()
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Pages/Charts.razor:346]: Rename to ChangeVerticalDispersionPlaneAsync

File: WoWsShipBuilder.Web/Pages/Charts.razor [hunk 10]
```diff
@@ -149,365 +150,310 @@
...
-        CreateBallisticChartDataset(name, data, ChartsHelper.BallisticParameter.ImpactVelocity, chartLineColor);
-        CreateBallisticChartDataset(name, data, ChartsHelper.BallisticParameter.ImpactAngle, chartLineColor);
+        SetProcessing(true);
+        List<IEnumerable<Point>> selectedVerticals = displayedShips.Select(item => ChartsHelper.SelectVerticalDispersionDataset(vertDispersionsCache[item.Key], selectedVertDispPlane)).ToList();
+        await BatchUpdateData(VerticalDispersionId, displayedShips.Keys.ToList(), selectedVerticals);
+        SetProcessing(false);
     }
 
-    private void CreateBallisticChartDataset(string name, Dictionary<double, Ballistic> data, ChartsHelper.BallisticParameter ballisticParameter, string chartLineColor)
+    private async Task UpdateTrajectory(double aimingRange)
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Pages/Charts.razor:354]: Rename to UpdateTrajectoryAsync

File: WoWsShipBuilder.Web/Pages/Charts.razor [hunk 11]
```diff
@@ -149,365 +150,310 @@
...
+        range = aimingRange;
+        List<IEnumerable<Point>> trajectoryData = displayedShips.Select(item => ChartsHelper.CreateTrajectoryDataset(ballisticCache[item.Key], range)).ToList();
+        await BatchUpdateData(TrajectoryId, displayedShips.Keys.ToList(), trajectoryData);
+        await ChangeSuggestedMax(TrajectoryId, aimingRange / 4 * 1000);
+        SetProcessing(false);
     }
-
-    private static void AddChartDataset(string name, IEnumerable<Point> data, LineConfig chartConfig, string chartLineColor)
+    
+    private async Task EditCharts()
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Pages/Charts.razor:364]: Rename to EditChartsAsync

File: WoWsShipBuilder.Web/Pages/Charts.razor [hunk 12]
```diff
@@ -149,365 +150,310 @@
...
+            return;
+        }
+        
+        await UpdateShipsBatch(((IEnumerable<ShellSelectionWrapper>) selectedShipList.Data).ToList());
+        SetProcessing(false);
     }
-
-    private void ReloadCharts()
+    
+    private async Task<DialogResult?> GetSelectedShips()
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Pages/Charts.razor:378]: Rename to GetSelectedShipsAsync

File: WoWsShipBuilder.Web/Pages/Charts.razor [hunk 13]
```diff
@@ -149,365 +150,310 @@
...
+            PenetrationId,
+            FlightTimeId,
+            ImpactVelocityId,
+            ImpactAngleId,
+        };
+        return chartIds;
     }
 
-    private async Task ApplyChartTooltipConfig()
+    #region JSInterop stuff
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Pages/Charts.razor:407]: Move entire JSInterop into dedicated service, remove regions

File: WoWsShipBuilder.Web/Pages/Charts.razor [hunk 14]
```diff
@@ -149,365 +150,310 @@
...
+
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Pages/Charts.razor:408]: Add Async suffix to all async method names

File: WoWsShipBuilder.Web/Pages/Charts.razor [hunk 15]
```diff
@@ -149,365 +150,310 @@
...
-            };
-            selectedShipList.Add(wrapper);
+            await module.InvokeVoidAsync("BatchUpdateData", chartId, labels, datas);
         }
-        else
+    }
+    
+    #endregion
+    
+    private void SetProcessing(bool newProcessing)
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Pages/Charts.razor:443]: Add explanation, why is this necessary? What does this do? Where is this used?

File: WoWsShipBuilder.Web/Pages/Charts.razor [hunk 16]
```diff
@@ -149,365 +150,310 @@
...
-
-    private string GetColor()
+        
+    async ValueTask IAsyncDisposable.DisposeAsync()
     {
-        if (colorCounter > ChartColors.Length - 1)
+        if (module is not null)
         {
-            colorCounter = 0;
+            await module.DisposeAsync();
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Pages/Charts.razor:456]: Catch exception when disposing. If DisposeAsync is called after the browser window is closed, this will always throw an exception.
Do not catch all exceptions, only catch relevant exceptions!

File: WoWsShipBuilder.Web/Utility/ChartsHelper.cs [hunk 17]
```diff
@@ -1,127 +1,19 @@
-﻿using ChartJs.Blazor.Common;
-using ChartJs.Blazor.Common.Axes;
-using ChartJs.Blazor.Common.Enums;
-using ChartJs.Blazor.LineChart;
-using ChartJs.Blazor.Util;
+﻿// using ChartJs.Blazor.Common;
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Utility/ChartsHelper.cs:1]: Remove commented code

File: WoWsShipBuilder.Web/Utility/ChartsHelper.cs [hunk 18]
```diff
@@ -1,127 +1,19 @@
...
+using System.Diagnostics.CodeAnalysis;
+using DynamicData;
 using WoWsShipBuilder.Core.DataContainers;
-using WoWsShipBuilder.Core.Localization;
-using WoWsShipBuilder.Core.Translations;
 using WoWsShipBuilder.DataStructures;
 
 namespace WoWsShipBuilder.Web.Utility;
 
 public static class ChartsHelper
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Utility/ChartsHelper.cs:15]: Add tests, consider to turn this into an injectable service to reduce coupling

File: WoWsShipBuilder.Web/Utility/ChartsHelper.cs [hunk 19]
```diff
@@ -210,14 +129,8 @@ public enum BallisticParameter
...
-        FlightTime,
-        ImpactVelocity,
-        ImpactAngle,
-        Trajectory,
-    }
 }
+
+// lowercase needed for chartjs to recognize to which axis the data belongs
+[SuppressMessage("ReSharper", "InconsistentNaming")]
+public record Point(double x, double y);
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Utility/ChartsHelper.cs:136]: If it's needed for JSInterop, mark the record as sealed

File: WoWsShipBuilder.Web/Utility/ShellSelectionWrapper.cs [hunk 20]
```diff
@@ -4,17 +4,28 @@ namespace WoWsShipBuilder.Web.Utility;
...
     }
 
     public string ShipIndex { get; }
 
-    public double MaxRange { get; set; } = -1;
+    public double MaxRange { get; set; }
 
     public Dictionary<string, ArtilleryData> ArtilleryDataDictionary { get; }
+
+    protected bool Equals(ShellSelectionWrapper other)
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Utility/ShellSelectionWrapper.cs:20]: Override actual Equals method and don't create a new custom method for equality check. If it's not supposed to be public, use a different name.

File: WoWsShipBuilder.Web/WoWsShipBuilder.Web.csproj [hunk 21]
```diff
@@ -75,4 +74,9 @@
       <AvaloniaResource Include="wwwroot\assets\nation_flags\flag_Usa.png" />
     </ItemGroup>
 
+    <ItemGroup>
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/WoWsShipBuilder.Web.csproj:77]: Remove this ItemGroup entirely

**Review by floribe2000 — CHANGES_REQUESTED** [review]
Refactor method and variable names, don't use abbreviations or acronyms when it's not necessary (disp instead of dispersion is not really helpful)
See individual comments for more details, not all naming issues have comments tho.
Also ensure to fix uncaught exceptions and infinite render loops.

File: WoWsShipBuilder.Web/Pages/Charts.razor (hunk 1 above)
Comment (yesod30) [reply WoWsShipBuilder.Web/Pages/Charts.razor:38]: No, chartjs specifically need a canvas with a set id. that is a variable simply so it's all centered in one location, without having to write it in every place,leading to possible typos. Ref is also not the same as id, since ref is an actual referece to the object, id is just an html tag.

File: WoWsShipBuilder.Web/Pages/Charts.razor (hunk 2 above)
Comment (yesod30) [reply WoWsShipBuilder.Web/Pages/Charts.razor:53]: Same as before

File: WoWsShipBuilder.Web/Pages/Charts.razor (hunk 3 above)
Comment (yesod30) [reply WoWsShipBuilder.Web/Pages/Charts.razor:81]: Same as before

File: WoWsShipBuilder.Web/Pages/Charts.razor (hunk 4 above)
Comment (yesod30) [reply WoWsShipBuilder.Web/Pages/Charts.razor:102]: Guess what, same as before

File: WoWsShipBuilder.Web/WoWsShipBuilder.Web.csproj (hunk 21 above)
Comment (yesod30) [reply WoWsShipBuilder.Web/WoWsShipBuilder.Web.csproj:77]: This is not related to this pull request, it probably got picked up from the blazor web app branch when i created this branch

File: WoWsShipBuilder.Web/Pages/Charts.razor (hunk 6 above)
Comment (yesod30) [reply WoWsShipBuilder.Web/Pages/Charts.razor:270]: Yes, because it lowers the amount of JSInterop exchange. i can put a comment with an explanation in the code too, but basically: IEnumerable<Point> is a single dataset. The middle List contains all datasets of a certain chart. The outermost List contains the list of datasets of each chart.

File: WoWsShipBuilder.Web/Utility/ShellSelectionWrapper.cs (hunk 20 above)
Comment (yesod30) [reply WoWsShipBuilder.Web/Utility/ShellSelectionWrapper.cs:20]: Removed because not used anymore

File: WoWsShipBuilder.Web/Services/ChartJsInterop.cs [hunk 22]
```diff
@@ -0,0 +1,76 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using Microsoft.JSInterop;
+using WoWsShipBuilder.Web.Utility;
+
+namespace WoWsShipBuilder.Web.Services;
+
+public class ChartJsInterop : IAsyncDisposable
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Services/ChartJsInterop.cs:7]: Not really sure about this class name. Probably fine tho

File: WoWsShipBuilder.Web/Utility/ShellSelectionWrapper.cs [hunk 23]
```diff
@@ -4,17 +4,19 @@ namespace WoWsShipBuilder.Web.Utility;
...
     public string ShipIndex { get; }
 
-    public double MaxRange { get; set; } = -1;
+    public double MaxRange { get; set; }
 
     public Dictionary<string, ArtilleryData> ArtilleryDataDictionary { get; }
+
 }
 
 public sealed record ArtilleryData(ArtilleryShell Shell, Dispersion DispersionValues, bool ShellSelected)
```
[the remark this answers is not in the document]
Comment (floribe2000) [reply WoWsShipBuilder.Web/Utility/ShellSelectionWrapper.cs:22]: Fix in feature/blazor-server-app

**Review by floribe2000 — APPROVED** [review]
LGTM
